examples/fluids/qfunctions/newtonian.h

```diff
@@ -254,54 +222,48 @@ CEED_QFUNCTION(IFunction_Newtonian)(void *ctx, CeedInt Q,
                                     const CeedScalar *const *in, CeedScalar *const *out) {
   // *INDENT-OFF*
   // Inputs
-  const CeedScalar (*q)[CEED_Q_VLA] = (const CeedScalar(*)[CEED_Q_VLA])in[0],
+  const CeedScalar (*q)[CEED_Q_VLA]         = (const CeedScalar(*)[CEED_Q_VLA])in[0],
                    (*Grad_q)[5][CEED_Q_VLA] = (const CeedScalar(*)[5][CEED_Q_VLA])in[1],
-                   (*q_dot)[CEED_Q_VLA] = (const CeedScalar(*)[CEED_Q_VLA])in[2],
-                   (*q_data)[CEED_Q_VLA] = (const CeedScalar(*)[CEED_Q_VLA])in[3],
-                   (*x)[CEED_Q_VLA] = (const CeedScalar(*)[CEED_Q_VLA])in[4];
+                   (*q_dot)[CEED_Q_VLA]     = (const CeedScalar(*)[CEED_Q_VLA])in[2],
+                   (*q_data)[CEED_Q_VLA]    = (const CeedScalar(*)[CEED_Q_VLA])in[3],
+                   (*x)[CEED_Q_VLA]         = (const CeedScalar(*)[CEED_Q_VLA])in[4];
   // Outputs
-  CeedScalar (*v)[CEED_Q_VLA] = (CeedScalar(*)[CEED_Q_VLA])out[0],
+  CeedScalar (*v)[CEED_Q_VLA]         = (CeedScalar(*)[CEED_Q_VLA])out[0],
              (*Grad_v)[5][CEED_Q_VLA] = (CeedScalar(*)[5][CEED_Q_VLA])out[1],
-             (*jac_data)[CEED_Q_VLA] = (CeedScalar(*)[CEED_Q_VLA])out[2];
+             (*jac_data)[CEED_Q_VLA]  = (CeedScalar(*)[CEED_Q_VLA])out[2];
   // *INDENT-ON*
   // Context
   NewtonianIdealGasContext context = (NewtonianIdealGasContext)ctx;
-  const CeedScalar *g     = context->g;
-  const CeedScalar dt     = context->dt;
+  const CeedScalar *g = context->g;
+  const CeedScalar dt = context->dt;
 
   CeedPragmaSIMD
   // Quadrature Point Loop
   for (CeedInt i=0; i<Q; i++) {
-    CeedScalar U[5];
-    for (CeedInt j=0; j<5; j++) U[j] = q[j][i];
+    CeedScalar qi[5];
+    for (CeedInt j=0; j<5; j++) qi[j] = q[j][i];
     const CeedScalar x_i[3] = {x[0][i], x[1][i], x[2][i]};
-    State s = StateFromU(context, U, x_i);
+    State s = StateFromQi(context, qi, x_i);
```

Comment:
An alternative is to make this function (what is now called `IFunction_Newtonian`) into a `CEED_QFUNCTION_HELPER` that takes as additional argument(s) the function pointers for the appropriate version of these operations. Then we'll have an `IFunction_Newtonian_Primitive` and `IFunction_Newtonian_Conservative` that are one-liners, just calling the helper with suitable extra args. This has a similar effect to C++ templates, but if you compile with optimization off, you can still step through the code you wrote in the debugging, including seeing/using the function pointers as defined. Since `CEED_QFUNCTION_HELPER` behaves like `static inline`, we should get both specializations optimized.

Reply:
I believe I have implemented this on the `jrwrigh/unite_fluids_pointers` branch, but I'm actually seeing practically identical results to what I was seeing before; the HIP implementation is about the same (or at least within noise of the measurement), but the CPU is consistently about 4% slower.

The branch currently keeps an `if (gas->primitive)` statement for handling the `U_dot` state, but manually removing that actually makes it a little bit slower (maybe 0.5%).

I'm running this on noether. Here's the `make info` for the compilation (in case that may be having an affect somehow):
```
CC            = gcc                                                                                                                                                                  
CXX           = g++                                                                                                                                                                  
FC            = gfortran                                                                                                                                                             
CPPFLAGS      = -I./include  -D__HIP_PLATFORM_HCC__ -D__HIP_PLATFORM_AMD__ -I/opt/rocm-5.1.3/hip/include -I/opt/rocm-5.1.3/llvm/bin/../lib/clang/14.0.0 -I/opt/rocm-5.1.3/hsa/include
CFLAGS        = $(OPT) $(CFLAGS.$(CC_VENDOR)) $(if $(ASAN),$(AFLAGS))                                                                                                                
CXXFLAGS      = $(OPT) $(CXXFLAGS.$(CC_VENDOR))                                                                                                                                      
FFLAGS        = $(OPT) $(FFLAGS.$(FC_VENDOR)) $(if $(ASAN),$(AFLAGS))                                                                                                                
NVCCFLAGS     = -ccbin $(CXX) -Xcompiler "$(OPT)" -Xcompiler -fPIC                                                                                                                   
HIPCCFLAGS    = $(filter-out $(OMP_SIMD_FLAG),$(OPT)) -fPIC -munsafe-fp-atomics                                                                                                      
CEED_LDFLAGS  = $(if $(ASAN),$(AFLAGS))                                                                                                                                              
CEED_LDLIBS   = -lm                                                                                                                                                                  
AR            = ar                                                                                                                                                                   
ARFLAGS       = crD                                                                                                                                                                  
OPT           = -O -g -march=native -ffp-contract=fast -fopenmp-simd                                                                                                                 
AFLAGS        = -fsanitize=address                                                                                                                                                   
```

Reply:
I tried with `OPT='-O3 -funsafe-math-optimizations -march=native'`, and the results are:

| Branch | Time (s) |
| ---- | ---- | 
| `main` | ~11.5 |
|  `jrwrigh/unite_fluids_primitive` (current) | ~11.7 |
|  `jrwrigh/unite_fluids_pointers` | ~15.3 |

So the pointers are not having the intended affect at all. I can only guess that they're not being inlined at all, though `-O3` should have the `-findirect-inlining` flag enabled through it. I tried removing the above mentioned if-statement, but the results are more-or-less the same.

Reply:
Alright, now I've got it.

A) Using `-fopt-info-inline` to dump out the optimization information, I confirmed that gcc was not inlining the helper function into the wrapper function _NOR_ was it inlining the `StateFrom*` function pointers into the helper function.

B) Prepending `__attribute__((always_inline))` to the beginning of the helper function forced gcc to inline it into the wrapper. Then gcc correctly inlined the `StateFrom*` into the wrapper function (ie. after the helper function was inlined). Doing that, I'm (finally) getting performance parity with the separate QFunctions.

Obviously the `__attribute__` is gcc only. I can't find an equivalent for intel compilers, though it's possible it might automatically inline the helper function into the wrapper. I'd like to know why gcc wasn't inlining the function even though we told it was available.

Reply:
>  I'd like to know why gcc wasn't inlining the function even though we told it was available.

Using `-Winline`, I get:
```
inlining failed in call to ‘IFunction_Newtonian_Generic’: --param large-stack-frame-growth limit reached [-Winline]
```

I was able to use `--param large-stack-frame=10000` to get the function to inline as expected. Can't say I really understand what that parameter is or how gcc uses it, but that's what was causing the inlining to fail.

Reply:
Thanks for working this out. Does `__attribute__((flatten))` applied to `CEED_QFUNCTION` achieve the same result?

We can use `always_inline`, but I'd like to make sure we still have good debugging info at low optimization levels. I see that both gcc and clang `#define __NO_INLINE__ 1` when compiling at `-O0` (default) and drop that, instead `#define __OPTIMIZE__ 1` when compiling with optimization. We could use this if needed to ensure that qfunction helpers are *not* inlined in debug builds, but are always inlined in optimized builds.

Reply:
`__attribute__((flatten))` works and actually makes it faster too. Going this route also ropes in intel compilers with `#pragma forceinline`.

> We can use always_inline, but I'd like to make sure we still have good debugging info at low optimization levels.

Yeah, I had the same thought. Would we want to prepend `__attribute__((flatten))` to the `CEED_QFUNCTION` definition depending on `__NO_INLINE__` and/or `__OPTIMIZE__`? Intel specifies a `__NO_INLINE__` macro, but not an `__OPTIMIZE__` macro.

Reply:
It's possible that the `__attribute__((flatten))` doesn't have an effect at `-O0` so please check that first. But if it still inlines, then yeah, I would
```c
#ifdef __NO_INLINE__ // and maybe compiler identifiers
#  define CEED_QFUNCTION_ATTR __attribute__((flatten))
#else
#  define CEED_QFUNCTION_ATTR
#endif
```
and then use `CEED_QFUNCTION_ATTR` in the `CEED_QFUNCTION()` macro.

Reply:
Yeah, `__attribute__((flatten))` does cause inlining at `-O0`.